# Introduction to US Federal Government Contracts

* *This notebook is part of the [Government Procurement Queries](https://github.com/antontarasenko/gpq) project*

## The Dataset

The BigQuery dataset (17 years of data, 45mn transactions, $6.7tn worth of goods and services):

- [gpqueries:contracts](https://bigquery.cloud.google.com/dataset/gpqueries:contracts)

*Important:* You need a Google account and a Google Cloud project to access the data (both free). Google offers you to create a new project when you open BigQuery. Do it. Then you'll need to follow Google's instructions and enable BigQuery in this project.

## Table `gpqueries:contracts.raw`

Table [`gpqueries:contracts.raw`](https://bigquery.cloud.google.com/table/gpqueries:contracts.raw) contains the unmodified data from the [USASpending.gov archives](https://www.usaspending.gov/DownloadCenter/Pages/dataarchives.aspx). It's constructed from `<year>_All_Contracts_Full_20160515.csv.zip` files and includes contracts from 2000 to May 15, 2016.

Table `gpqueries:contracts.raw` contains 45M rows and 225 columns.

Each row refers to a transaction (a purchase or refund) made by a federal agency. It may be a pizza or an airplane.

The columns are grouped into categories:

- Transaction: `unique_transaction_id`-`baseandalloptionsvalue`
- Buyer (government agency): `maj_agency_cat`-`fundedbyforeignentity`
- Dates: `signeddate`-`lastdatetoorder`, `last_modified_date`
- Contract: `contractactiontype`-`programacronym`
- Contractor (supplier, vendor): `vendorname`-`statecode`
- Place of performance: `PlaceofPerformanceCity`-`placeofperformancecongressionaldistrict`
- Product or service bought: `psc_cat`-`manufacturingorganizationtype`
- General contract information: `agencyid`-`idvmodificationnumber`
- Competitive procedure: `solicitationid`-`statutoryexceptiontofairopportunity`
- Contractor details: `organizationaltype`-`otherstatutoryauthority`
- Contractor's executives: `prime_awardee_executive1`-`interagencycontractingauthority`

Detailed description for each variable is available in the official codebook:

- [`USAspending.govDownloadsDataDictionary.pdf`](https://www.usaspending.gov/DownloadCenter/Documents/USAspending.govDownloadsDataDictionary.pdf)

## Queries

### BigQuery Web GUI

You can execute queries mentioned here at <https://bigquery.cloud.google.com/table/gpqueries:contracts.raw> (press "Compose query").

### Datalab

This notebook was written in Google Datalab. You may need the libraries imported below to replicate it:

In [1]:
import gcp.bigquery as bq

Within Datalab, you can define queries with [`sql` magic](https://github.com/catherinedevlin/ipython-sql) like this:

In [2]:
%%sql --module gpq

define query totals
select
  count(*) transactions,
  sum(dollarsobligated) sum_dollarsobligated,
  count(unique(dunsnumber)) vendors,
  count(unique(solicitationid)) purchase_procedures
from 
  gpqueries:contracts.raw

And execute with `bq` to get a dataframe:

In [3]:
bq.Query(gpq.totals).to_dataframe()

transactions  sum_dollarsobligated  vendors  purchase_procedures
0      44500941          6.653578e+12   621532              2223674

Which means we're dealing with 44.5M transactions totalling 6.7 trillion dollars. These purchases came from 622k vendors that won 2.2mn solicitations issued by government agencies.

## Data Mining Government Clients

Suppose you want to start selling to the government. While [FBO.gov](http://www.fbo.gov/) publishes government RFPs and you can apply there, government agencies often issue requests when they've already chosen the supplier. Agencies go through FBO.gov because it's a mandatory step for deals north of $25K. But winning at this stage is unlikely if an RFP is already tailored for another supplier.

Reaching warm leads in advance would increase chances of winning a government contract. The contracts data helps identify the warm leads by looking at purchases in the previous years.

There're several ways of searching through those years.

### Who Buys What You Make

The goods and services bought in each transaction are encoded in the variable `productorservicecode`. Top ten product categories according to this variable:

In [4]:
%%sql
select
  substr(productorservicecode, 1, 4) product_id,
  first(substr(productorservicecode, 7)) product_name,
  count(*) transactions,
  sum(dollarsobligated) sum_dollarsobligated
from
  gpqueries:contracts.raw
group by
  product_id
order by
  sum_dollarsobligated desc
limit 10

QueryResultsTable job_5OYeRQNV79dkoo0LffBT7Ba6N2Y

You can find agencies that buy products like yours. If it's "software":

In [5]:
%%sql
select
  substr(agencyid, 1, 4) agency_id,
  first(substr(agencyid, 7)) agency_name,
  count(*) transactions,
  sum(dollarsobligated) sum_dollarsobligated
from
  gpqueries:contracts.raw
where
  productorservicecode contains 'software'
group by
  agency_id
order by
  sum_dollarsobligated desc
ignore case

QueryResultsTable job_1jg_ZQ5tFKg8pHFGX6fnuYGiDaY

### What Firms in Your Industry Sell to the Government

Another way to find customers is the variable `principalnaicscode` that encodes the industry in which the vendor does business.

The list of NAICS codes is available at [Census.gov](http://www.census.gov/cgi-bin/sssd/naics/naicsrch?chart=2012), but you can do text search in the table. Let's find who bought software from distributors in 2015:

In [6]:
%%sql
select
  substr(agencyid, 1, 4) agency_id,
  first(substr(agencyid, 7)) agency_name,
  substr(principalnaicscode, 1, 6) naics_id,
  first(substr(principalnaicscode, 9)) naics_name,
  count(*) transactions,
  sum(dollarsobligated) sum_dollarsobligated
from
  gpqueries:contracts.raw
where
  principalnaicscode contains 'software' and
  fiscal_year = 2015
group by
  agency_id, naics_id
order by
  sum_dollarsobligated desc
ignore case

QueryResultsTable job_tyO5qqBhadXXwwH8p2u9qTutzoA

### Inspecting Specific Transactions

You can learn details from looking at transactions for a specific `(agency, NAICS)` pair. For example, what software does TSA buy?

In [7]:
%%sql
select
  fiscal_year,
  dollarsobligated,
  vendorname, city, state, annualrevenue, numberofemployees,
  descriptionofcontractrequirement
from
  gpqueries:contracts.raw
where
  agencyid contains 'transportation security administration' and
  principalnaicscode contains 'computer and software stores'
ignore case

QueryResultsTable job_16eFmMJh3axW3zdfW7ioeWdHVE0

Alternatively, specify vendors your product relates to and check how the government uses it. Top deals in data analytics:

In [8]:
%%sql
select
  agencyid,
  dollarsobligated,
  vendorname,
  descriptionofcontractrequirement
from
  gpqueries:contracts.raw
where
  vendorname contains 'tableau' or
  vendorname contains 'socrata' or
  vendorname contains 'palantir' or
  vendorname contains 'revolution analytics' or
  vendorname contains 'mathworks' or
  vendorname contains 'statacorp' or
  vendorname contains 'mathworks'
order by
  dollarsobligated desc
limit
  100
ignore case

QueryResultsTable job_GHwVcGuA_3SqI6q4FiMMI4k9yqY

### Searching Through Descriptions

Full-text search and regular expressions for the variable `descriptionofcontractrequirement` narrow results for relevant product groups:

In [9]:
%%sql
select
  agencyid,
  dollarsobligated,
  descriptionofcontractrequirement
from
  gpqueries:contracts.raw
where
  descriptionofcontractrequirement contains 'body camera'
limit
  100
ignore case

QueryResultsTable job_wIUPYZekHe4jwd-oHsy76OSUD4c

Some rows of `descriptionofcontractrequirement` contain codes like "IGF::CT::IGF". These codes classify the purchase into three groups of "[Inherently Governmental Functions](https://www.fpds.gov/fpdsng_cms/index.php/en/newsroom/108-nherently-governmental-functions.html)" (IGF):

1. IGF::CT::IGF for Critical Functions
2. IGF::CL::IGF for Closely Associated
3. IGF::OT::IGF for Other Functions

### Narrowing Your Geography

You can find local opportunities using variables for vendors (`city`, `state`) and services sold (`PlaceofPerformanceCity`, `pop_state_code`). The states where most contracts are delivered in:

In [10]:
%%sql
select
  substr(pop_state_code, 1, 2) state_code,
  first(substr(pop_state_code, 4)) state_name,
  sum(dollarsobligated) sum_dollarsobligated
from
  gpqueries:contracts.raw
group by
  state_code
order by
  sum_dollarsobligated desc

QueryResultsTable job_xz8OgI40jd2GPlDk9J5fqZjxTDY

## Facts about Government Contracting

Let's check some popular statements about government contracting.

### Small Businesses Win Most Contracts

Contractors had to report their revenue and the number of employees. It makes easy to check if small business is welcomed in government contracting:

In [11]:
%%sql --module gpq

define query vendor_size_by_agency
select
  substr(agencyid, 1, 4) agency_id,
  first(substr(agencyid, 7)) agency_name,
  nth(11, quantiles(annualrevenue, 21)) vendor_median_annualrevenue,
  nth(11, quantiles(numberofemployees, 21)) vendor_median_numberofemployees,
  count(*) transactions,
  sum(dollarsobligated) sum_dollarsobligated
from
  gpqueries:contracts.raw
group by
  agency_id
having
  transactions > 1000 and
  sum_dollarsobligated > 10e6
order by
  vendor_median_annualrevenue asc

In [12]:
bq.Query(gpq.vendor_size_by_agency).to_dataframe()

agency_id                                        agency_name  \
0        1242         ANIMAL AND PLANT HEALTH INSPECTION SERVICE   
1        1528             IMMIGRATION AND NATURALIZATION SERVICE   
2        5800                FEDERAL EMERGENCY MANAGEMENT AGENCY   
3        1900                               STATE, DEPARTMENT OF   
4        9568                    BROADCASTING BOARD OF GOVERNORS   
5        7200               AGENCY FOR INTERNATIONAL DEVELOPMENT   
6        2800                     SOCIAL SECURITY ADMINISTRATION   
7        1226                                     FOREST SERVICE   
8        12E3                              RURAL HOUSING SERVICE   
9        2028                                U.S. SECRET SERVICE   
10       2001                               DEPARTMENTAL OFFICES   
11       4500            EQUAL EMPLOYMENT OPPORTUNITY COMMISSION   
12       9524                           NATIONAL MEDIATION BOARD   
13       1560              ATF ACQUISITION AND PROPERTY MGMT DIV   
14       1227                          SOIL CONSERVATION SERVICE   
15       2026                FINANCIAL CRIME ENFORCEMENT NETWORK   
16       2022           ALCOHOL AND TOBACCO TAX AND TRADE BUREAU   
17       12C2                                     FOREST SERVICE   
18       7520                              PUBLIC HEALTH SERVICE   
19       6950                          UNITED STATES COAST GUARD   
20       1422                          BUREAU OF LAND MANAGEMENT   
21       2024            FEDERAL LAW ENFORCEMENT TRAINING CENTER   
22       6100                 CONSUMER PRODUCT SAFETY COMMISSION   
23       1438  OFFICE OF SURFACE MINING, RECLAMATION AND ENFO...   
24       1153         UNITED STATES TRADE AND DEVELOPMENT AGENCY   
25       1230                      AGRICULTURAL RESEARCH SERVICE   
26       2046          OFFICE OF THE COMPTROLLER OF THE CURRENCY   
27       1603                            OFFICE OF THE SOLICITOR   
28       4740                           PUBLIC BUILDINGS SERVICE   
29       12C3             NATURAL RESOURCES CONSERVATION SERVICE   
..        ...                                                ...   
121      2044                                 UNITED STATES MINT   
122      2700                  FEDERAL COMMUNICATIONS COMMISSION   
123      9507               COMMODITY FUTURES TRADING COMMISSION   
124      4705                  OFFICE OF THE ADMINISTRATOR(ACMD)   
125      7529                      NATIONAL INSTITUTES OF HEALTH   
126      1540                              FEDERAL PRISON SYSTEM   
127      1665               PENSION BENEFIT GUARANTY CORPORATION   
128      7003          U.S. CITIZENSHIP AND IMMIGRATION SERVICES   
129      1344                        PATENT AND TRADEMARK OFFICE   
130      3355                            NATIONAL GALLERY OF ART   
131      6000                          RAILROAD RETIREMENT BOARD   
132      19BM  INTERNATIONAL BOUNDARY AND WATER COMMISSION: U...   
133      8961               FEDERAL ENERGY REGULATORY COMMISSION   
134      4704                        OFFICE OF INSPECTOR GENERAL   
135      3600                    VETERANS AFFAIRS, DEPARTMENT OF   
136      4732                        FEDERAL ACQUISITION SERVICE   
137      8000      NATIONAL AERONAUTICS AND SPACE ADMINISTRATION   
138      1549                    FEDERAL BUREAU OF INVESTIGATION   
139      12D2                                FARM SERVICE AGENCY   
140      6940     NATIONAL HIGHWAY TRAFFIC SAFETY ADMINISTRATION   
141      6930                    FEDERAL RAILROAD ADMINISTRATION   
142      0559                    GAO, EXCEPT COMPTROLLER GENERAL   
143      7001                   OFFICE OF PROCUREMENT OPERATIONS   
144      7505  OFFICE OF ASSISTANT SECRETARY FOR ADMINISTRATI...   
145      9000                           SELECTIVE SERVICE SYSTEM   
146      7528         AGENCY FOR HEALTHCARE RESEARCH AND QUALITY   
147      1436     BUREAU OF SAFETY AND ENVIRONMENTAL ENFORCE

The median shows the most likely supplier. Agencies on the top of the table actively employ vendors whose annual revenue is less than $1mn.

The Department of Defence, the largest buyer with $4.5tn worth of goods and services bought over these 17 years, has the median vendor with $2.5mn in revenue and 20 employees. It means that half of the DoD's vendors have less than $2.5mn in revenue.

### Set-Aside Deals Take a Small Share

Set-aside purchases are reserved for special categories of suppliers, like women-, minority-, and veteran-owned businesses. There's a lot of confusion about their share in transactions. We can settle this confusion with data:

In [13]:
%%sql
select
  womenownedflag,
  count(*) transactions,
  sum(dollarsobligated) sum_dollarsobligated
from
  gpqueries:contracts.raw
group by
  womenownedflag

QueryResultsTable job_iOFnq8ieeeHX2gxQWTCCr4ZOEuE

Women-owned businesses make about one tenth of the transactions, but their share in terms of sales is only 3.7%.

A cross-tabulation for major set-aside categories:

In [14]:
%%sql
select
  womenownedflag, veteranownedflag, minorityownedbusinessflag,
  count(*) transactions,
  sum(dollarsobligated) sum_dollarsobligated
from
  gpqueries:contracts.raw
group by
  womenownedflag, veteranownedflag, minorityownedbusinessflag
order by
  womenownedflag, veteranownedflag, minorityownedbusinessflag desc

QueryResultsTable job_Qh4DhmZZecFIWeF7Wlah0cQwfSM

For example, firms owned by women, veterans, and minorities (all represented at the same time) sell $5bn in goods and services. That's 0.07% of all government purchases.

### New Vendors Emerge Each Year

Becoming a government contractor may seem difficult at first, but let's see how many new contractors the government had in 2015.

In [15]:
%%sql
select
  sum(if(before2015.dunsnumber is null, 1, 0)) new_vendors,
  sum(if(before2015.dunsnumber is null, 0, 1)) old_vendors
from
  flatten((select unique(dunsnumber) dunsnumber from gpqueries:contracts.raw where fiscal_year = 2015), dunsnumber) in2015
left join
  flatten((select unique(dunsnumber) dunsnumber from gpqueries:contracts.raw where fiscal_year < 2015), dunsnumber) before2015
  on before2015.dunsnumber = in2015.dunsnumber

QueryResultsTable job_OjFASl4YsdPFMqmG63vz8lXqgzc

(The variable `dunsnumber` refers to the unique number each contractor obtains at [SAM.gov](https://www.sam.gov/portal/SAM/). It identifies unique contractors better than `vendorname`, which varies.)

In 2015, 12% of the suppliers turned out to be companies that had never sold to the government.

## More

This was a short intro to government contracting data.

For more, check the [Government Procurement Queries](https://github.com/antontarasenko/gpq) project.

## Contacts

Anton Tarasenko

<mailto:antontarasenko@gmail.com>

<http://antontarasenko.com>

